File: Python/Data Structures & Algorithms Module - Python/Week6/ArrayListStack.py
"""
myArrayStack.py
Billy Cussen
16/11/2020
"""

class ArrayListStack:
    def __init__(self):
        self.myArrayList = []

    def getSize(self):
        return len(self.myArrayList)

    def isEmpty(self):
        return self.getSize() == 0

    def toString(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't traverse!")
        position = 0
        index = self.getSize()-1
        result = "***ARRAY STACK TRAVERSAL***\n"
        return self.toStringHelper(self.myArrayList,result,position, index)

    def toStringHelper(self, curr, result, position, index):
        if index == -1:
            result+="***END***\n"
            return result
        result += "Position "+str(position)+", Element "+str(curr[position])+"\n"
        position+=1
        index-=1
        return self.toStringHelper(curr,result,position, index)

    def push(self,elem):
        if self.isEmpty():
            self.myArrayList.append(elem)
        else:
            self.myArrayList.insert(0,elem)

    def pop(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't pop!")
        else:
            elem = self.myArrayList[0]
            self.myArrayList.pop(0)
            return elem

    def peek(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't peek!")
        else:
            return self.myArrayList[0]

myArrayStack = ArrayListStack()

try:
    print("Is Empty: "+str(myArrayStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myArrayStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myArrayStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myArrayStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myArrayStack.toString())
                except Exception as e:
                    print(str(e)+"\n")

myArrayStack.push(2)
myArrayStack.push(4)
myArrayStack.push(6)
myArrayStack.push(8)
myArrayStack.push(10)
myArrayStack.push(12)

try:
    print("Is Empty: "+str(myArrayStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myArrayStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myArrayStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myArrayStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myArrayStack.toString())
                except Exception as e:
                    print(e)

while myArrayStack.isEmpty() == False:
    print("Pop: "+str(myArrayStack.pop()))

try:
    print("Is Empty: "+str(myArrayStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myArrayStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myArrayStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myArrayStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myArrayStack.toString())
                except Exception as e:
                    print(e)

File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week6/ArrayListStack.java
/*
ArrayListStack.java
Billy Cussen
16/11/2020
*/

import java.util.List;
import java.util.ArrayList;

public class ArrayListStack<T>{

    List<T> myList;

    ArrayListStack(List<T> myList){
        this.myList = myList;
    }

    public static void main(String [] args){

        List<Integer> myList = new ArrayList<Integer>();
        ArrayListStack<Integer> myArrayListStack = new ArrayListStack<Integer>(myList);

        try{
            System.out.println("Is Empty: "+myArrayListStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myArrayListStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myArrayListStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myArrayListStack.toStringArrayListStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage()+"\n");
                    } 
                }    
            }    
        }

        myArrayListStack.push(2);
        myArrayListStack.push(4);
        myArrayListStack.push(6);
        myArrayListStack.push(8);
        myArrayListStack.push(10);

        try{
            System.out.println("Is Empty: "+myArrayListStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myArrayListStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myArrayListStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myArrayListStack.toStringArrayListStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage());
                    } 
                }    
            }    
        }
        
        while(!myArrayListStack.isEmpty()){
            System.out.println(myArrayListStack.pop());
        }

        try{
            System.out.println("Is Empty: "+myArrayListStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myArrayListStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myArrayListStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myArrayListStack.toStringArrayListStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage());
                    } 
                }    
            }    
        }

    }

    void push(T element){
        if(isEmpty()){
            this.myList.add(element);
        } else{
            this.myList.add(0, element);
        }
    }

    T pop(){
        if (isEmpty()){
            throw new IndexOutOfBoundsException("Can't Pop, Stack is Empty!");
        } else{
            T element = this.myList.get(0);
            this.myList.remove(0);
            return element;
        }
    }

    T peek(){
        if (isEmpty()){
            throw new IndexOutOfBoundsException("Can't peek, Stack is Empty!");
        } else{
            return this.myList.get(0);
        }
    }

    int getSize(){
        int size = 0;
        for(T e: this.myList){
            if (e != null){
                size++;
            }
        }
        return size;
    }

    boolean isEmpty(){
        return getSize() == 0;
    }

    String toStringArrayListStack(){
        if(isEmpty()){
            throw new IndexOutOfBoundsException("Can't traverse, Stack is Empty!");
        }
        String result = "***ARRAY STACK TRAVERSAL***\n";
        int position = 0;
        return toStringHelper(this.myList,result, position);
    }

    String toStringHelper(List<T> current, String result, int position){
        if(isEmpty()){
            result+="***END***\n";
            return result;
        }
        result+="Position "+position+", Element: "+current.get(0)+"\n";
        current.remove(0);
        return toStringHelper(current, result, ++position);
    }
}
